DataFormats/TrackingRecHitSoA/interface/TrackingRecHitsDevice.h

```diff
@@ -9,51 +9,66 @@
-  using PortableDeviceCollection<TrackingRecHitLayout<TrackerTraits>, TDev>::buffer;
-
-  TrackingRecHitDevice(edm::Uninitialized)
-      : PortableDeviceCollection<TrackingRecHitLayout<TrackerTraits>, TDev>{edm::kUninitialized} {}
-
-  // Constructor which specifies the SoA size, number of BPIX1 hits, and the modules entry points
-  template <typename TQueue>
-  explicit TrackingRecHitDevice(TQueue queue, uint32_t nHits, int32_t offsetBPIX2, uint32_t const* hitsModuleStart)
-      : PortableDeviceCollection<TrackingRecHitLayout<TrackerTraits>, TDev>(nHits, queue), offsetBPIX2_{offsetBPIX2} {
-    auto start_h = cms::alpakatools::make_device_view(queue, hitsModuleStart, TrackerTraits::numberOfModules + 1);
-    auto start_d =
-        cms::alpakatools::make_device_view(queue, view().hitsModuleStart().data(), TrackerTraits::numberOfModules + 1);
-    alpaka::memcpy(queue, start_d, start_h);
-
-    auto off_h = cms::alpakatools::make_host_view(offsetBPIX2_);
-    auto off_d = cms::alpakatools::make_device_view(queue, view().offsetBPIX2());
-    alpaka::memcpy(queue, off_d, off_h);
-  }
-
-  uint32_t nHits() const { return view().metadata().size(); }
-
-  int32_t offsetBPIX2() const { return offsetBPIX2_; }
-
-  uint32_t const* hitsModuleStart() const { return view().hitsModuleStart().data(); }
-
-  // asynchronously update the information cached within the class itself from the information on the device
-  template <typename TQueue>
-  void updateFromDevice(TQueue queue) {
-    auto off_h = cms::alpakatools::make_host_view(offsetBPIX2_);
-    auto off_d = cms::alpakatools::make_device_view(queue, view().offsetBPIX2());
-    alpaka::memcpy(queue, off_h, off_d);
-  }
-
-private:
-  // offsetBPIX2 is used on host functions so is useful to have it also stored in the class and not only in the layout
-  int32_t offsetBPIX2_ = 0;
-};
+// TODO: The class is created via inheritance of the PortableCollection.
+// This is generally discouraged, and should be done via composition.
+// See: https://github.com/cms-sw/cmssw/pull/40465#discussion_r1067364306
+
+namespace reco {
+
+  template <typename TDev>
+  using HitPortableCollectionDevice = PortableDeviceMultiCollection<TDev, reco::TrackingRecHitSoA, reco::HitModuleSoA>;
+
+  template <typename TDev>
+  class TrackingRecHitDevice : public HitPortableCollectionDevice<TDev> {
+  public:
+    TrackingRecHitDevice() = default;
+
+    TrackingRecHitDevice(edm::Uninitialized) : HitPortableCollectionDevice<TDev>{edm::kUninitialized} {}
+
+    // Constructor which specifies only the SoA size, to be used when copying the results from host to device
+    template <typename TQueue>
+    explicit TrackingRecHitDevice(TQueue queue, uint32_t nHits, uint32_t nModules)
+        : HitPortableCollectionDevice<TDev>({{int(nHits), int(nModules)}}, queue) {}
+
+    // Constructor from clusters
+    template <typename TQueue>
+    explicit TrackingRecHitDevice(TQueue queue, SiPixelClustersDevice<TDev> const &clusters)
```

Comment:
Not a review comment, but - could you remind me where this constructor is used ?

Reply:
When building the hits [here](https://github.com/cms-sw/cmssw/blob/17f86b8f731bd6adc6bd36bb1b5442a6f95abaf1/RecoLocalTracker/SiPixelRecHits/plugins/alpaka/PixelRecHitKernels.dev.cc#L43).